Add support for trame_jupyter_extension (trame over jupyter comm)

alesgenova opened the pull request:
### Overview

Add support in pyvista for the new [`trame_jupyter_extension`](https://github.com/Kitware/trame-jupyter-extension)

When this extension is installed (and the flag added in this PR is `True`), the trame server and client will communicate through the existing jupyter comm, instead of creating a new websocket connection.


### Notes
- I have removed the launching of the server from the `set_jupyter_backend()` function, relying on the fact that the server will be started in `show_trame()` if it's not running. The reason why I did it (other than the fact that it seemed redundant) is because if the server is prematurely started with `aiohttp` wslink backend, it will not be possible to make it work with the `jupyter` backend, and vice versa. So it seemed like a good idea to delay the starting of the server as much as possible until we actually need to display something. While this seemed like a reasonable change to me, let me know if there were deeper reasons for it to be there.
- It seemed that both the vue2 and vue3 viewers were being embedded in a vue2 `VAppLayout`. According to @jourdain this shouldn't be happening, so I changed the `BaseViewer` class so that each subclass is in charge of creating the appropriate layout.

banesullivan commented on pyvista/trame/ui/base_viewer.py:
```diff
@@ -258,3 +259,21 @@ def export(self):  # numpydoc ignore=RT01
         """Implement in derived classes."""
         raise NotImplementedError()
+
+    def make_layout(self, *args, **kwargs) -> AbstractLayout:
```
```suggestion
    def make_layout(self, *args, **kwargs) -> AbstractLayout:  # pragma: no cover
```

banesullivan reviewed (changes requested):
Looking great and working excellently! Some minor feedback left inline. Would you also please update `doc/source/user-guide/jupyter/trame.rst` with usage instruction for this and any potential tradeoffs or benefits for using this?

Also, would you please comment on how this extension may or may not help with #4708, and #4652? I think this will resolve those issues.

banesullivan commented:
Re the server launching changes you mentioned:

> While this seemed like a reasonable change to me, let me know if there were deeper reasons for it to be there.

I think you took the right approach here, thanks for cleaning this up.

> It seemed that both the vue2 and vue3 viewers were being embedded in a vue2 VAppLayout

Thanks for cleaning this up. Seems like it was an artifact of #4811

alesgenova commented:
> @alesgenova, let me know how I can support you as this is the final PR we'd like to sneak in for the `0.43` release in the coming days (see #5138)

Thanks for the review, it won't be a problem to address the minor issues :)
I'm planning on having it done on Monday, unless it's too late for the new release?

banesullivan commented:
That should work! Thanks!

alesgenova commented on pyvista/plotting/themes.py:
```diff
@@ -1469,6 +1479,33 @@ def server_proxy_prefix(self):  # numpydoc ignore=RT01
+        """Return whether the trame_jupyter_extension is detected."""
+        return self._jupyter_extension_available
+
+    @jupyter_extension_available.setter
+    def jupyter_extension_available(self, _available: bool):  # numpydoc ignore=GL08
+        warnings.warn(
+            "The jupyter_extension_available flag is read only and is automatically detected."
+        )
+
```
Originally I omitted the setter like you suggest here, however one of the unit tests was failing because the setter was missing, and that's why I ended up adding a noop setter.

Agreed that the theme might not be the best place for available flag. If you have a better place in mind, I'll be happy to move it there

banesullivan replied:
Ah, I didn't realize a test was complaining about this. I think let's just leave this as is then. Thanks for looking into this!

alesgenova commented:
@banesullivan 
I addressed all the comments from your review (minus the `extension_available` setter, see my comment above), you can find the changes in the last commit added to the PR.

I also tested out the example reported in #4652 , and can confirm that at least locally it works when using the new extension, while it's broken (as expected) when using the server proxy. Not definitive but seems promising.

banesullivan commented:
Looks fantastic! Thank you for addressing all of my feedback. One remaining question, though: is there any disadvantage to using this jupyter comms infrastructure here? I believe at some point @jourdain and I thought it may be limited compared to the capabilities of wslink

banesullivan commented:
Not sure if this needs to be included in the docs, but going to add it as a comment here for posterity and future readers: if running on a _Jupyter**Hub**_, you will need to install either of these (`jupyter_server_proxy` or `trame_jupyter_extension`) in the base Python environment where the JupyterHub server is running.

banesullivan reviewed (dismissed):
LGTM! Happy to give the approval and merge regardless of the testing coverage check. IMO, testing is sufficient here.

tkoyama010 commented on pyvista/trame/ui/base_viewer.py:
```diff
@@ -258,3 +259,21 @@ def export(self):  # numpydoc ignore=RT01
         raise NotImplementedError()
+
+    def make_layout(self, *args, **kwargs) -> AbstractLayout:  # pragma: no cover
+        """Create instance of an AbstractLayout which is appropriate for a concrete viewer.
```
```suggestion
        """Create an instance of an AbstractLayout which is appropriate for a concrete viewer.
```

tkoyama010 commented on pyvista/trame/jupyter.py:
```diff
@@ -149,6 +168,8 @@ def launch_server(server=None, port=None, host=None, **kwargs):
     if host is None:
         # Default to `127.0.0.1` unless user sets TRAME_DEFAULT_HOST
         host = os.environ.get('TRAME_DEFAULT_HOST', '127.0.0.1')
+    if wslink_backend is None and pyvista.global_theme.trame.jupyter_extension_enabled:
+        wslink_backend = "jupyter"
```
```suggestion
    if wslink_backend is None and pyvista.global_theme.trame.jupyter_extension_enabled:  # pragma: no cover
        wslink_backend = "jupyter"
```

tkoyama010 commented on pyvista/trame/jupyter.py:
```diff
@@ -334,12 +361,19 @@ def handler(viewer, src, **kwargs):
         server = get_server(name=pyvista.global_theme.trame.jupyter_server_name)
     else:
         server = get_server(name=name)
     if name is None and not server.running:
```
```suggestion
    if name is None and not server.running:  # pragma: no cover
```

tkoyama010 commented on pyvista/trame/jupyter.py:
```diff
@@ -356,19 +390,27 @@ def handler(viewer, src, **kwargs):
-        host=kwargs.get('host', 'localhost'),
-        protocol=kwargs.get('protocol', 'http'),
-    )
+    if jupyter_extension_enabled:
```
```suggestion
    if jupyter_extension_enabled:  # pragma: no cover
```

tkoyama010 reviewed (dismissed):
LGTM. Great contribution!

tkoyama010 commented on pyvista/trame/jupyter.py:
```diff
@@ -334,12 +361,19 @@ def handler(viewer, src, **kwargs):
         server = get_server(name=name)
-    if name is None and not server.running:
-        elegantly_launch(server)
+    if name is None and not server.running:  # pragma: no cover
```
```suggestion
    if name is None and not server.running:
```

tkoyama010 commented on pyvista/trame/jupyter.py:
```diff
@@ -334,12 +361,19 @@ def handler(viewer, src, **kwargs):
-        elegantly_launch(server)
+    if name is None and not server.running:  # pragma: no cover
+        wslink_backend = "aiohttp"
+        if jupyter_extension_enabled:
```
```suggestion
        if jupyter_extension_enabled: # pragma: no cover
```

tkoyama010 commented:
pre-commit.ci autofix

banesullivan commented:
Sorry I jumped the gun a bit there, @tkoyama010! I think we're ready to start shipping the release so I merged this

tkoyama010 commented:
No problem. We waited enough.

jourdain commented:
> I believe at some point @jourdain and I thought it may be limited compared to the capabilities of wslink

The binary messages remain binary, so we should be good. The overhead is most likely less than the server-proxy.

jourdain commented:
yes for jupyter definitely
